include/snap/plugin.h

```diff
@@ -23,43 +23,121 @@ limitations under the License.
 
+/**
+ * Type is the plugin type
+ */
 enum Type {
-  collector,
-  processor,
-  publisher
+  Collector,
+  Processor,
+  Publisher
 };
 
-class PluginInterface {
+/**
+ * RpcType is the rpc protocol the plugin should use.
+ * Only GRPC is supported for C++.
+ */
+enum RpcType {
+  GRPC = 2,
+};
 
+/**
+ * Strategy is the routing and caching strategy this plugin requires.
+ * A routing and caching Strategy is the method which snapd uses to cache 
+ * metrics, and select the correct running instance of a plugin.
+ */
+enum Strategy {
+  /**
+   * LRU: Least Recently Used.
+   * The default strategy.
+   */
+  LRU,
+
+  /**
+   * Sticky: Sticky Running plugins have a 1:1 relationship to Tasks.
+   */
+  Sticky,
+
+  /** ConfigBased: Bases routing decisions on the incoming config.
+   * The config based strategy hashes the incoming config data and
+   * uses it as a key to select a running plugin.  It can be useful for plugins
+   * which maintain a connection to a database, for instance.
+   */
+  ConfigBased
+};
+
+/**
+ * Meta is the metadata about the plugin.
+ */
+struct Meta final {
   public:
-    virtual Config::Policy getConfigPolicy() = 0;
+    Meta(Type type, std::string name, int version) : type(type),
+                                                     name(name),
+                                                     version(version) {};
+
+    Type type;
+    std::string name;
+    int version;
+
+    // These members all have defaults.
+    RpcType rpc_type = RpcType::GRPC;
+    int concurrency_count = 5;
+    bool exclusive = false;
+    bool unsecure = false;
+    std::chrono::milliseconds cache_ttl = std::chrono::milliseconds(500);
+    Strategy strategy = Strategy::LRU;
 };
 
-class CollectorInterface : public PluginInterface {
+/**
+ * PluginInterface is the interface implemented by ALL plugins.
+ * Every plugin must implement get_config_policy.
+ */
+class PluginInterface {
+  public:
+    virtual Config::Policy get_config_policy() = 0;
+};
 
+/**
+ * The interface for a collector plugin.
+ * A Collector the source.
+ * It is responsible for collecting metrics in the snap pipeline.
+ */
+class CollectorInterface : public PluginInterface {
   public:
-    virtual std::vector<Metric::Metric> getMetricTypes(Config::Config
-                                                             cfg) = 0;
 
+    /*
+     * get_metric_types should report all the metrics this plugin can collect.
+     */
     virtual std::vector<Metric::Metric>
-    collectMetrics(std::vector<Metric::Metric> metrics) = 0;
+    get_metric_types(Config::Config cfg) = 0;
+
+    /*
+     * collect_metrics is given a list of metrics to collect.
+     * It should collect and annotate each metric with the apropos context.
+     */
+    virtual void collect_metrics(std::vector<Metric::Metric>* metrics) = 0;
 };
 
+/**
+ * The interface for a Processor plugin.
+ * A Processor is a filter, or it may _decorate_ the data as it passes through
+ * the pipeline.
```

Comment:
I think maybe this definition of a processor should be opened up. A processor can do anything with the metrics including filter, decorate and derive.